SingularityService/src/main/java/com/hubspot/singularity/helpers/TaskLagGuardrail.java

```diff
@@ -0,0 +1,57 @@
+package com.hubspot.singularity.helpers;
+
+import com.google.inject.Inject;
+import com.hubspot.singularity.SingularityPendingTaskId;
+import com.hubspot.singularity.config.SingularityConfiguration;
+import com.hubspot.singularity.data.TaskManager;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+public class TaskLagGuardrail {
+  private final TaskManager taskManager;
+  private final SingularityConfiguration configuration;
+  private ConcurrentMap<String, Integer> lateTasksByRequestId;
+  private long lastUpdate;
+
+  @Inject
+  public TaskLagGuardrail(
+    SingularityConfiguration configuration,
+    TaskManager taskManager
+  ) {
+    this.configuration = configuration;
+    this.taskManager = taskManager;
+    this.lateTasksByRequestId = new ConcurrentHashMap<>();
+    this.lastUpdate = 0;
+  }
+
+  public boolean isLagged(String requestId) {
+    updateLateTasksByRequestId();
+    return lateTasksByRequestId.containsKey(requestId);
+  }
+
+  private void updateLateTasksByRequestId() {
+    long now = System.currentTimeMillis();
+    if (now - lastUpdate > 1000L * configuration.getRequestCacheTtl()) {
```

Comment:
Do we need to care about request cache ttl at all here? Shouldn't this be running on the leader only?

Reply:
This was mostly me not wanting to add another configuration option, and thinking that request cache ttl was an appropriate value for this.